lib/rules/no-constant-binary-expression.js

```diff
@@ -0,0 +1,460 @@
+    }
+}
+
+/**
+ * Test if an AST node will always result in a newly constructed object
+ * @param {Scope} scope The scope in which the node was found.
+ * @param {ASTNode} node The node to test
+ * @returns {boolean} Will `node` always be new?
+ */
+function isAlwaysNew(scope, node) {
+    switch (node.type) {
+        case "ObjectExpression":
+        case "ArrayExpression":
+        case "ArrowFunctionExpression":
+        case "FunctionExpression":
+        case "ClassExpression":
+            return true;
+        case "NewExpression": {
+            if (node.callee.type !== "Identifier") {
+                return false;
+            }
+
+            /*
+             * All the built-in constructors are always new, but
+             * user-defined constructors could return a sentinel
+             * object.
+             *
+             * Catching these is especially useful for primitive constructures
+             * which return boxed values, a surprising gotcha' in JavaScript.
+             */
+            return Object.hasOwnProperty.call(globals.builtin, node.callee.name) &&
+              isReferenceToGlobalVariable(scope, node.callee);
+        }
+        case "Literal":
+
+            // Regular expressions are objects, and thus always new
+            return typeof node.regex === "object";
+        case "SequenceExpression": {
+            const last = node.expressions[node.expressions.length - 1];
+
+            return isAlwaysNew(scope, last);
+        }
+        case "AssignmentExpression":
+            if (node.operator === "=") {
+                return isAlwaysNew(scope, node.right);
+            }
+            return false;
+        case "ConditionalExpression":
+            return isAlwaysNew(scope, node.consequent) && isAlwaysNew(scope, node.alternate);
+        case "JSXElement": // ESLint has a policy of not assuming any specific JSX behavior.
+        case "JSXFragment":
+            return false;
+        default:
+            return false;
+    }
+}
+
+
+/**
+ * Checks if one operand will cause the result to be constant.
+ * @param {Scope} scope Scope in which the expression was found.
+ * @param {ASTNode} a One side of the expression
+ * @param {ASTNode} b The other side of the expression
+ * @param {string} operator The binary expression operator
+ * @returns {ASTNode | null} The node which will cause the expression to have a constant result.
+ */
+function findBinaryExpressionConstantOperand(scope, a, b, operator) {
+    if (operator === "==" || operator === "!=") {
+        if (
+            (isNullOrUndefined(a) && hasConstantNullishness(scope, b)) ||
+            (isStaticBoolean(scope, a) && hasConstantLooseBooleanComparison(scope, b)) ||
+
+            /*
+             * If both sides are "new", then both sides are objects and
+             * therefore they will be compared by reference even with `==`
+             * equality.
+             */
+            (isAlwaysNew(scope, a) && isAlwaysNew(scope, b))
```

Comment:
I can't figure out why this check logically belongs to `findBinaryExpressionConstantOperand`, while the similar one for `===` and `!==` is done separately? Their messages are also different in the current version of this PR:

```js
/* eslint no-constant-binary-expression: "error" */

[a] === [b]; // Unexpected comparison to newly constructed object. These two values can never be equal.

[a] == [b]; // Unexpected constant binary expression. Compares constantly with the left-hand side of the `==`.

```

Reply:
Good point. I've pulled this out and given it its own message.